pykokkos/lib/ufuncs.py

```diff
@@ -157,11 +169,24 @@ def log2(view):
         Output view.
 
     """
-    if str(view.dtype) == "DataType.double":
-        pk.parallel_for(view.shape[0], log2_impl_1d_double, view=view)
-    elif str(view.dtype) == "DataType.float":
-        pk.parallel_for(view.shape[0], log2_impl_1d_float, view=view)
-    return view
+    out = pk.View(view.shape, view.dtype)
+    if "double" in str(view.dtype) or "float64" in str(view.dtype):
+        if view.shape == ():
+            # NOTE: is this really worth sending to a kernel?
+            pk.parallel_for(1, log2_impl_1d_double, view=view, out=out)
```

Comment:
For another ufunc that Umair worked on (I forget which one exactly), I asked him to simply call the `math.func()` on the scalar. You could do that here too if you think it's better.

Reply:
`math.log2(view)` doesn't work here because it raises `ValueError: math domain error` when the input is `0`. I think NumPy returns `-inf` and only raises a warning. Perhaps C++/kokkos are similar to NumPy in that regard.

One other thing to think about is 0-d vs. scalar input, and if we want to return a "Python scalar" vs. a "pykokkos scalar." And what would a "pykokkos scalar" even mean? NumPy scalars can be confusing, they often print like regular Python scalars, but they are not the same thing (for example, a NumPy scalar integer wouldn't have the arbitrary precision/size of a pure Python integer [bignums]).

Reply:
Does the array API say anything about 0-d and scalar inputs?